File: ui/article-renderer/src/main/java/org/cru/godtools/article/fragment/ArticlesFragment.java
package org.cru.godtools.article.fragment;

import android.app.Application;
import android.arch.lifecycle.AndroidViewModel;
import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.ViewModelProviders;
import android.os.Bundle;
import android.support.annotation.MainThread;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import org.cru.godtools.article.R;
import org.cru.godtools.article.R2;
import org.cru.godtools.article.adapter.ArticlesAdapter;
import org.cru.godtools.articles.aem.db.ManifestAssociationRepository;
import org.cru.godtools.articles.aem.model.Article;
import org.cru.godtools.base.tool.fragment.BaseToolFragment;

import java.util.List;
import java.util.Locale;

import butterknife.BindView;
import timber.log.Timber;

public class ArticlesFragment extends BaseToolFragment implements ArticlesAdapter.Callback {
    public static final String TAG = "ArticlesFragment";
    private static final String MANIFEST_KEY = "manifest-key";

    @Nullable
    @BindView(R2.id.articles_recycler_view)
    RecyclerView mArticlesView;
    @Nullable
    ArticlesAdapter mArticlesAdapter;

    String mManifestKey = "";

    // these properties should be treated as final and only set/modified in onCreate()
    @NonNull
    private /*final*/ ArticleListViewModel mViewModel;

    public static ArticlesFragment newInstance(@NonNull final String code, @NonNull final Locale locale,
                                               String manifestKey) {
        final ArticlesFragment fragment = new ArticlesFragment();
        final Bundle args = new Bundle();
        populateArgs(args, code, locale);
        args.putString(MANIFEST_KEY, manifestKey);
        fragment.setArguments(args);
        return fragment;
    }

    // region LifeCycle Events

    @Override
    public void onCreate(@Nullable final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        final Bundle args = getArguments();
        if (args != null) {
            mManifestKey = args.getString(MANIFEST_KEY, mManifestKey);
        }

        mViewModel = ViewModelProviders.of(this).get(ArticleListViewModel.class);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_articles, container, false);
    }

    @Override
    public void onViewCreated(@NonNull final View view, @Nullable final Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        setupArticlesView();
    }

    /**
     * This is the Override Method from BaseToolFragment. This is fired anytime the Manifest Object is updated and will
     * update the adapter of it's changes.
     */
    @Override
    protected void onManifestUpdated() {
        super.onManifestUpdated();
        updateArticlesViewManifest();
    }

    /**
     * This is the callback method from ArticleAdapter that will handle the functionality of an article being selected
     * from the list.
     *
     * @param article the selected Article
     */
    @Override
    public void onArticleSelected(@Nullable final Article article) {
        Timber.tag(TAG).d("You selected \"%s\" as your article", article.mTitle);
    }

    @Override
    public void onDestroyView() {
        cleanupArticlesView();
        super.onDestroyView();
    }

    // endregion LifeCycle Events

    // region ArticlesView

    /**
     * This method will initialize all of the required data for the RecyclerView
     */
    private void setupArticlesView() {
        if (mArticlesView != null) {
            mArticlesView.addItemDecoration(
                    new DividerItemDecoration(mArticlesView.getContext(), DividerItemDecoration.VERTICAL));

            mArticlesAdapter = new ArticlesAdapter();
            mArticlesAdapter.setCallbacks(this);
            mViewModel.getArticles(mManifestKey).observe(this, mArticlesAdapter);
            mArticlesView.setAdapter(mArticlesAdapter);

            updateArticlesViewManifest();
        }
    }

    private void updateArticlesViewManifest() {
        if (mArticlesAdapter != null) {
            mArticlesAdapter.setToolManifest(mManifest);
        }
    }

    private void cleanupArticlesView() {
        if (mArticlesAdapter != null) {
            mArticlesAdapter.setCallbacks(null);
            mViewModel.getArticles(mManifestKey).removeObserver(mArticlesAdapter);
        }
        mArticlesAdapter = null;
    }

    // endregion ArticlesView

    public static class ArticleListViewModel extends AndroidViewModel {
        @Nullable
        private LiveData<List<Article>> mArticles;

        public ArticleListViewModel(@NonNull final Application application) {
            super(application);
        }

        @NonNull
        @MainThread
        LiveData<List<Article>> getArticles(@Nullable final String manifestKey) {
            if (mArticles == null) {
                mArticles = new ManifestAssociationRepository(getApplication()).getArticlesByManifestID(manifestKey);
            }

            return mArticles;
        }
    }
}

File: ui/article-renderer/src/main/java/org/cru/godtools/article/ui/articles/ArticlesFragment.kt
package org.cru.godtools.article.ui.articles

import android.os.Bundle
import androidx.fragment.app.viewModels
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.observe
import androidx.recyclerview.widget.DividerItemDecoration
import java.util.Locale
import javax.inject.Inject
import org.ccci.gto.android.common.androidx.lifecycle.combineWith
import org.ccci.gto.android.common.androidx.lifecycle.emptyLiveData
import org.ccci.gto.android.common.androidx.lifecycle.switchCombineWith
import org.ccci.gto.android.common.util.MainThreadExecutor
import org.ccci.gto.android.common.util.WeakTask
import org.ccci.gto.android.common.util.findListener
import org.cru.godtools.article.R
import org.cru.godtools.article.aem.db.ArticleDao
import org.cru.godtools.article.aem.model.Article
import org.cru.godtools.article.aem.service.AemArticleManager
import org.cru.godtools.article.databinding.FragmentArticlesBinding
import org.cru.godtools.base.tool.fragment.BaseToolFragment
import org.cru.godtools.base.tool.service.ManifestManager
import org.cru.godtools.base.tool.viewmodel.LatestPublishedManifestDataModel
import splitties.fragmentargs.argOrNull

private val syncCompleteTask = WeakTask.Task<MutableLiveData<Boolean>> { it.value = false }

class ArticlesFragment : BaseToolFragment<FragmentArticlesBinding>, ArticlesAdapter.Callbacks {
    constructor() : super(R.layout.fragment_articles)
    constructor(code: String, locale: Locale, category: String? = null) :
        super(R.layout.fragment_articles, code, locale) {
        this.category = category
    }

    interface Callbacks {
        fun onArticleSelected(article: Article?)
    }

    private var category by argOrNull<String>()

    // region Lifecycle
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setupDataModel()
    }

    override fun onBindingCreated(binding: FragmentArticlesBinding, savedInstanceState: Bundle?) {
        super.onBindingCreated(binding, savedInstanceState)
        binding.manifest = toolDataModel.manifest
        binding.setupArticlesView()
        binding.setupSwipeRefresh()
    }

    /**
     * This is the callback method from ArticleAdapter that will handle the functionality of an article being selected
     * from the list.
     *
     * @param article the selected Article
     */
    override fun onArticleSelected(article: Article?) {
        findListener<Callbacks>()?.onArticleSelected(article)
    }
    // endregion Lifecycle

    // region Data Model
    override val toolDataModel: ArticlesFragmentDataModel by viewModels()

    private fun setupDataModel() {
        toolDataModel.category.value = category
    }
    // endregion Data Model

    @Inject
    internal lateinit var aemArticleManager: AemArticleManager

    private val isSyncing = MutableLiveData(false)
    private fun syncData(force: Boolean) {
        val result = aemArticleManager.enqueueSyncManifestAemImports(toolDataModel.manifest.value, force)
        isSyncing.value = true
        result.addListener(WeakTask(isSyncing, syncCompleteTask), MainThreadExecutor())
    }

    // region View Logic
    // region ArticlesView
    private val articlesAdapter by lazy {
        ArticlesAdapter(this, toolDataModel.manifest).also {
            it.callbacks.set(this)
            toolDataModel.articles.observe(this, it)
        }
    }

    private fun FragmentArticlesBinding.setupArticlesView() {
        articles.apply {
            addItemDecoration(DividerItemDecoration(context, DividerItemDecoration.VERTICAL))
            adapter = articlesAdapter
        }
    }
    // endregion ArticlesView

    private fun FragmentArticlesBinding.setupSwipeRefresh() {
        refresh.setOnRefreshListener { syncData(true) }
        isSyncing.observe(viewLifecycleOwner) { refresh.isRefreshing = it }
    }
    // endregion View Logic
}

class ArticlesFragmentDataModel @Inject constructor(
    manifestManager: ManifestManager,
    private val articleDao: ArticleDao
) : LatestPublishedManifestDataModel(manifestManager) {
    internal val category = MutableLiveData<String?>()

    private val tags = manifest.combineWith(category) { manifest, category ->
        when (category) {
            null -> null
            else -> manifest?.findCategory(category)?.aemTags.orEmpty()
        }
    }

    internal val articles =
        toolCode.switchCombineWith(locale, tags) { tool, locale, tags ->
        when {
            tool == null || locale == null -> emptyLiveData<List<Article>>()
            tags == null -> articleDao.getArticles(tool, locale)
            else -> articleDao.getArticles(tool, locale, tags.toList())
        }
    }
}
